### PR TITLE
Cr 1113771 -  Add command line auto-completion to xbutil and xbmgmt

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
[https://jira.xilinx.com/browse/CR-1113771](url)
xbutil and xbmgmt sub-commands becomes longer than before. To help users use these utilities, please add second level sub-commands auto-completion feature for bash.

#### How problem was solved, alternative solutions (if any) and why they were rejected
Added Bash and tcsh autocomplete scripts in to the codebase. The setup scripts were changed to source the added scripts to enable the autocomplete features.

#### Risks (if any) associated the changes in the commit
Changes to the setup script may cause issues on some systems if they have the PATH set to previous installations of XRT and run the setup scripts. This could potential offer commands that are not available to the older version of XRT. This can be mitigated by recommending all users to source the setup scripts (as they should).

#### What has been tested and how, request additional testing if necessary
tcsh example
```
xsjpranjal01:~% source /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt/setup.csh
/usr/local/bin/mis_env: No such file or directory.
Autocomplete enabled for the xbutil command
Autocomplete enabled for the xbmgmt command
XILINX_XRT        : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt/
PATH              : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//bin:/usr/sbin:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//bin:/usr/sbin:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//bin:/usr/sbin:/usr/sbin:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//bin:/usr/sbin:/opt/xilinx/xrt/bin:/usr/sbin:/usr/sbin:/opt/xilinx/xrt/bin:/usr/sbin:/home/dbenusov/bin:/usr/local/bin:/mis/TREE/bin:/usr/bin:/bin:/usr/ucb
LD_LIBRARY_PATH   : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//lib:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//lib:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//lib:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//lib:/opt/xilinx/xrt/lib:/opt/xilinx/xrt/lib:/usr/local/lib:/usr/lib
PYTHONPATH        : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//python:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//python:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//python:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//python:/opt/xilinx/xrt/python:/opt/xilinx/xrt/python
xsjpranjal01:~% xbmgmt
--batch   examine   -h        program   --verbose
dump      --force   --help    reset     --version
xsjpranjal01:~% xbmgmt program
--base             --force            --revert-to-golden --verbose
--batch            -h                 --shell            --version
-d                 --help             -u
--device           --image            --user
xsjpranjal01:~% xbmgmt program -
--base             --force            --revert-to-golden --verbose
--batch            -h                 --shell            --version
-d                 --help             -u
--device           --image            --user
xsjpranjal01:~% xbutil
configure examine   program   reset     validate
xsjpranjal01:~% xbutil
configure examine   program   reset     validate
xsjpranjal01:~% xbutil program
--batch   --device  -h        -u        --verbose
-d        --force   --help    --user    --version
xsjpranjal01:~% xbutil program --
--batch   --device  --force   --help    --user    --verbose --version
xsjpranjal01:~% xbutil program --ver
--verbose --version
xsjpranjal01:~% xbutil program --version
--batch   --device  -h        -u        --verbose
-d        --force   --help    --user    --version
xsjpranjal01:~% xbutil program --version -
```
bash example
```
dbenusov@xsjpranjal01:~$ source /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt/setup.sh
Autocomplete enabled for the xbutil command
Autocomplete enabled for the xbmgmt command
XILINX_XRT        : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt
PATH              : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt/bin:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//bin:/usr/sbin:/home/dbenusov/bin:/usr/local/bin:/mis/TREE/bin:/usr/bin:/bin:/usr/ucb
LD_LIBRARY_PATH   : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt/lib:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//lib:/usr/local/lib:/usr/lib
PYTHONPATH        : /proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt/python:/proj/xsjhdstaff6/dbenusov/XRT/build/Debug/opt/xilinx/xrt//python
dbenusov@xsjpranjal01:~$ xbutil
program    examine    reset      --batch    --help     --version
validate   configure  --verbose  --force    -h
dbenusov@xsjpranjal01:~$ xbutil validate -
--run      --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbutil validate -
--run      --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbutil validate --
--run      --output   --verbose  --force    --version
--format   --device   --batch    --help
dbenusov@xsjpranjal01:~$ xbutil validate -
--run      --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbutil validate --
--run      --output   --verbose  --force    --version
--format   --device   --batch    --help
dbenusov@xsjpranjal01:~$ xbutil validate --ver
--verbose  --version
dbenusov@xsjpranjal01:~$ xbutil validate --ver
--verbose  --version
dbenusov@xsjpranjal01:~$ xbutil validate --version -
--run      --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbmgmt
dump       program    --verbose  --force    -h
examine    reset      --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbmgmt examine -
--report   --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbmgmt examine -
--report   --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbmgmt examine --
--report   --output   --verbose  --force    --version
--format   --device   --batch    --help
dbenusov@xsjpranjal01:~$ xbmgmt examine --for
--format  --force
dbenusov@xsjpranjal01:~$ xbmgmt examine --for
--format  --force
dbenusov@xsjpranjal01:~$ xbmgmt examine --format asdf -
--report   --format   --output   --device   --verbose  --force    -h
-r         -f         -o         -d         --batch    --help     --version
dbenusov@xsjpranjal01:~$ xbmgmt examine --format asdf -
```
#### Documentation impact (if any)
Would need to write into the documentation that this is available.

### CODE REVIEW

**Comment by dayeh-xilinx** [comment]
retest this please. Looks like an infra error.

**Comment by xilinxgitops** [comment]
retest this looks like infra error.

File: src/runtime_src/core/tools/CMakeLists.txt [hunk 1]
```diff
@@ -11,3 +11,16 @@ if (${XRT_NATIVE_BUILD} STREQUAL "yes")
...
+    # https://cmake.org/cmake/help/latest/command/install.html#files
+    # This is used for the setup script to enable auto completion
+    install (PROGRAMS "./xbmgmt2/xbmgmt-bash-completion" DESTINATION "${XRT_INSTALL_DIR}/share/completions")
+    install (PROGRAMS "./xbmgmt2/xbmgmt-csh-completion" DESTINATION "${XRT_INSTALL_DIR}/share/completions")
+    install (PROGRAMS "./xbmgmt2/xbmgmt-csh-completion-wrapper" DESTINATION "${XRT_INSTALL_DIR}/share/completions")
+    install (PROGRAMS "./xbutil2/xbutil-bash-completion" DESTINATION "${XRT_INSTALL_DIR}/share/completions")
+    install (PROGRAMS "./xbutil2/xbutil-csh-completion" DESTINATION "${XRT_INSTALL_DIR}/share/completions")
+    install (PROGRAMS "./xbutil2/xbutil-csh-completion-wrapper" DESTINATION "${XRT_INSTALL_DIR}/share/completions")
+  endif()
+endif()
```
Comment (AShivangi) [inline src/runtime_src/core/tools/CMakeLists.txt:26]: Add an empty line after this

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# bash completion for xbmgmt                              -*- shell-script -*-
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2019-2021 Xilinx, Inc. All rights reserved.
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion:3]:
```suggestion
# Copyright (C) 2021 Xilinx, Inc. All rights reserved.
```

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+          ;;
+      esac
+      ;;
+  esac
+}
+
+complete  -o nosort -F _xbmgmt_completion xbmgmt
+echo Autocomplete enabled for the xbmgmt command
+
+# ex: filetype=sh
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-bash-completion:104]: Add empty line

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-csh-completion [hunk 4]
```diff
@@ -0,0 +1,100 @@
+#!/bin/csh -f
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2019-2021 Xilinx, Inc. All rights reserved.
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-csh-completion:3]:
```suggestion
# Copyright (C) 2021 Xilinx, Inc. All rights reserved.
```

File: src/runtime_src/core/tools/xbmgmt2/xbmgmt-csh-completion-wrapper [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/bin/csh -f
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2019-2021 Xilinx, Inc. All rights reserved.
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbmgmt2/xbmgmt-csh-completion-wrapper:3]:
```suggestion
# Copyright (C) 2021 Xilinx, Inc. All rights reserved.
```

File: src/runtime_src/core/tools/xbutil2/xbutil-bash-completion [hunk 6]
```diff
@@ -0,0 +1,110 @@
+# bash completion for xbutil                              -*- shell-script -*-
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2019-2021 Xilinx, Inc. All rights reserved.
```
Comment (AShivangi) [inline src/runtime_src/core/tools/xbutil2/xbutil-bash-completion:3]:
```suggestion
# Copyright (C) 2021 Xilinx, Inc. All rights reserved.
```

**Review by AShivangi — CHANGES_REQUESTED** [review]
Minor changes requested, but the code looks good overall! Thanks for adding extensive comments :) 

@stsoe, should we advertise this in the changelog?

**Comment by keryell** [comment]
Interesting feature! :-)
Are there equivalent features for, say, Vitis or Vivado?

**Review by stsoe — APPROVED** [review]
Interesting indeed.  I was not aware of this mechanism.  Doesn't look very scalable, but cool indeed.   

@keryell, no Vitis and Vivado does not have this at the command line level, but does have same completion (libedit) at the Tcl shell level at least.

**Review by AShivangi — APPROVED** [review]
Looks good!

**Review by rozumx — APPROVED** [review]
Good good.
Thank you for making these changes.

**Comment by dbenusov-xilinx** [comment]
retest this please.

**Comment by xilinxgitops** [comment]
retest this please

**Comment by sonals** [comment]
Simply awesome! It was on my personal wish list for so long...

**Pull request merged**